code/modules/urist/modules/GOAI/integrations/blocker_definitions/_shared.dm

```diff
@@ -4,12 +4,79 @@
 	var/atom/atom_owner = owner
 	var/datum/directional_blocker/blocker = null
 
+	if(atom_owner?.density)
+		blocker = new(ALL_CARDINAL_DIRS, TRUE, TRUE)
+	else
+		atom_owner?.blocker_gen_enabled = FALSE
+
+	return blocker
+
+/proc/GenerateGenericDirBlocker(var/owner = null, var/invert_dir = FALSE)
+	var/atom/atom_owner = owner
+	var/datum/directional_blocker/blocker = null
+
+	if(atom_owner?.density)
+		var/dir_to_block = invert_dir ? dir2opposite(atom_owner.dir) : atom_owner.dir
+		blocker = new(dir_to_block, FALSE, TRUE)
+	else
+		atom_owner?.blocker_gen_enabled = FALSE
+
+	return blocker
+
+
+/proc/GenerateDynamicFullBlocker(var/owner)
+	var/atom/atom_owner = owner
+	var/datum/directional_blocker/blocker = null
+
 	if(atom_owner)
-		if(atom_owner.density)
-			blocker = new(ALL_CARDINAL_DIRS, TRUE, TRUE)
+		blocker = new(ALL_CARDINAL_DIRS, TRUE, atom_owner.density)
+
+	//Density might change later on, so we cannot cache blocker_gen_enabled. Generate a blocker and check for updates when needed
+
+	return blocker
+
+/proc/GenerateDynamicDirBlocker(var/owner, var/invert_dir = FALSE)
+	var/atom/atom_owner = owner
+	var/datum/directional_blocker/blocker = null
 
-		else
-			// Caching to speed up queries
-			atom_owner.cover_gen_enabled = FALSE
+	if(atom_owner)
+		var/dir_to_block = invert_dir ? dir2opposite(atom_owner.dir) : atom_owner
```

Comment:
I think this is an oversight. This currently causes runtimes.
```suggestion
		var/dir_to_block = invert_dir ? dir2opposite(atom_owner.dir) : atom_owner.dir
```

Reply:
WHOOPS. Yep, that's totally a mistake on my side. I quickly changed to code to allow for inverting directions and obviously didn't read over it through enough. My bad.